### ISSUE TITLE
A socket timeout should NOT be enforced by a Shadowsocks server or client

### ISSUE THREAD

**Issue by database64128** [shadowsocks/shadowsocks-org#164]
In the [official JSON configuration scheme](https://shadowsocks.org/en/config/quick-guide.html), a `timeout` field is defined as a _connection timeout in seconds_. However, different implementations have different interpretations of the `timeout` field.

- `shadowsocks-libev`: It's interpreted as a socket timeout and was originally enforced both server-side and client-side. Client-side enforcement was later removed in shadowsocks/shadowsocks-libev@7ad2df9. Server-side default socket timeout has been discussed and changed multiple times.
- `shadowsocks-rust`: It's interpreted as a socket timeout and enforced both server-side and client-side.
- `shadowsocks-windows`: Connection timeout. Defaults to 5 seconds.
- `go2-shadowsocks`: Was never implemented, as intended. shadowsocks/go-shadowsocks2#94
- Other implementations and protocols: In v2ray-core, clash, outline-ss-server, trojan-go, the implementation of the Shadowsocks protocol has no timeout options. Other protocols of similar purposes have no such options.

The `timeout` field never actually served its original purpose. Instead, it causes problems for users when interpreted as a socket timeout, and basically serves no practical purpose when interpreted as a connection timeout. New implementations never implemented the `timeout` field.

In conclusion, it is my opinion that we should make the following changes:

- No socket timeout should be enforced by a Shadowsocks server or client.
- The `timeout` field should be removed from the official JSON configuration scheme. All client implementations should remove the related options. Server implementations may optionally provide an option for socket timeout, if the maintainer deems it necessary that such rare and extreme conditions are considered.

**Comment by Mygod** [comment]
@madeye could probably clarify on this but timeout could be used for TCP write timeout. For UDP, hole punching timeout is a commonly used practice.

**Comment by madeye** [comment]
Let's remove `timeout` from the default config. Any client and server can still support `timeout` as an optional feature.

**Comment by Mygod** [comment]
Yeah or we could just do that. 😅
